# California housing dataset

We'll be re-doing the example from chapter 2 of _Hands-on machine learning with Scikit-Learn, Keras & TensorFlow_. The text is heavily derived from that book and included here purely for didactive purposes.

## Loading the data

The data is already in our root directory. We'll use the `readCsv` function to read the data into a dataframe.

In [1]:
import qualified DataFrame as D

df <- D.readCsv "./housing.csv"

## Take a Quick Look at the Data Structure

Let’s take a look at the top five rows using the DataFrame’s `take` function.

In [2]:
D.take 5 df

--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
index<br>Int | longitude<br>Double | latitude<br>Double | housing_median_age<br>Double | total_rooms<br>Double | total_bedrooms<br>Maybe Double | population<br>Double | households<br>Double | median_income<br>Double | median_house_value<br>Double | ocean_proximity<br>Text
-------------|---------------------|--------------------|------------------------------|-----------------------|--------------------------------|----------------------|----------------------|-------------------------|------------------------------|------------------------
0            | -122.23             | 37.88              | 41.0                         | 880.0                 | Just 129.0                     | 322.0                | 126.0                | 8.3252                  | 452600.0                     | NEAR BAY               
1            | -122.22             | 37.86              | 21.0                         | 7099.0                | Just 1106.0                    | 2401.0               | 1138.0               | 8.3014                  | 358500.0                     | NEAR BAY               
2            | -122.24             | 37.85              | 52.0                         | 1467.0                | Just 190.0                     | 496.0                | 177.0                | 7.2574                  | 352100.0                     | NEAR BAY               
3            | -122.25             | 37.85              | 52.0                         | 1274.0                | Just 235.0                     | 558.0                | 219.0                | 5.6431000000000004      | 341300.0                     | NEAR BAY               
4            | -122.25             | 37.85              | 52.0                         | 1627.0                | Just 280.0                     | 565.0                | 259.0                | 3.8462                  | 342200.0                     | NEAR BAY               


Each row represents one district. There are 10 attributes: _longitude_, _latitude_, _housing_median_age_, total_rooms, _total_bedrooms_, _population_, _households_, _median_income_, _median_house_value_, and _ocean_proximity_.

We can already tell from the types that _total_bedrooms_ will have null values since it's type is `Maybe Double`. How many nulls does it have? We can get a summary of all rows using the `describeColumns` function.

In [3]:
D.describeColumns df

--------------------------------------------------------------------------------------------------------------------------------------------------------
index<br>Int | Column Name<br>Text | # Non-null Values<br>Int | # Null Values<br>Int | # Partially parsed<br>Int | # Unique Values<br>Int | Type<br>Text
-------------|---------------------|--------------------------|----------------------|---------------------------|------------------------|-------------
0            | total_bedrooms      | 20433                    | 207                  | 0                         | 1924                   | Maybe Double
1            | ocean_proximity     | 20640                    | 0                    | 0                         | 5                      | Text        
2            | median_house_value  | 20640                    | 0                    | 0                         | 3842                   | Double      
3            | median_income       | 20640                    | 0                    | 0                         | 12928                  | Double      
4            | households          | 20640                    | 0                    | 0                         | 1815                   | Double      
5            | population          | 20640                    | 0                    | 0                         | 3888                   | Double      
6            | total_rooms         | 20640                    | 0                    | 0                         | 5926                   | Double      
7            | housing_median_age  | 20640                    | 0                    | 0                         | 52                     | Double      
8            | latitude            | 20640                    | 0                    | 0                         | 862                    | Double      
9            | longitude           | 20640                    | 0                    | 0                         | 844                    | Double      


There are 20,640 instances in the dataset, which means that it is fairly small by
Machine Learning standards, but it’s perfect to get started. Notice that the total_bed
rooms attribute has only 20,433 non-null values, meaning that 207 districts are miss‐
ing this feature. We will need to take care of this later.

_ocean_proximity_ has only 5 unique values. Let's look at what those values are:

In [4]:
import Data.Text (Text)

D.frequencies ("ocean_proximity" :: ) df

-------------------------------------------------------------------------------------------------------------------------
index<br>Int | Statistic<br>Text | <1H OCEAN<br>Any | INLAND<br>Any | ISLAND<br>Any | NEAR BAY<br>Any | NEAR OCEAN<br>Any
-------------|-------------------|------------------|---------------|---------------|-----------------|------------------
0            | Count             | 9136             | 6551          | 5             | 2290            | 2658             
1            | Percentage (%)    | 44.26%           | 31.74%        | 0.02%         | 11.09%          | 12.88%           


Let’s look at the other fields. The `summarize` function shows a summary of the
numerical attributes

In [5]:
D.summarize df

--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
index<br>Int | Statistic<br>Text | longitude<br>Double | latitude<br>Double | housing_median_age<br>Double | total_rooms<br>Double | total_bedrooms<br>Double | population<br>Double | households<br>Double | median_income<br>Double | median_house_value<br>Double
-------------|-------------------|---------------------|--------------------|------------------------------|-----------------------|--------------------------|----------------------|----------------------|-------------------------|-----------------------------
0            | Count             | 20640.0             | 20640.0            | 20640.0                      | 20640.0               | 20433.0                  | 20640.0              | 20640.0              | 20640.0                 | 20640.0                     
1            | Mean              | -119.57             | 35.63              | 28.64                        | 2635.76               | 537.87                   | 1425.48              | 499.54               | 3.87                    | 206855.82                   
2            | Minimum           | -124.35             | 32.54              | 1.0                          | 2.0                   | 1.0                      | 3.0                  | 1.0                  | 0.5                     | 14999.0                     
3            | 25%               | -121.8              | 33.93              | 18.0                         | 1447.75               | 296.0                    | 787.0                | 280.0                | 2.56                    | 119600.0                    
4            | Median            | -118.49             | 34.26              | 29.0                         | 2127.0                | 435.0                    | 1166.0               | 409.0                | 3.53                    | 179700.0                    
5            | 75%               | -118.01             | 37.71              | 37.0                         | 3148.0                | 647.0                    | 1725.0               | 605.0                | 4.74                    | 264725.0                    
6            | Max               | -114.31             | 41.95              | 52.0                         | 39320.0               | 6445.0                   | 35682.0              | 6082.0               | 15.0                    | 500001.0                    
7            | StdDev            | 2.0                 | 2.14               | 12.59                        | 2181.62               | 421.39                   | 1132.46              | 382.33               | 1.9                     | 115395.62                   
8            | IQR               | 3.79                | 3.78               | 19.0                         | 1700.25               | 351.0                    | 938.0                | 325.0                | 2.18                    | 145125.0                    
9            | Skewness          | -0.3                | 0.47               | 6.0e-2                       | 4.15                  | 3.46                     | 4.94                 | 3.41                 | 1.65                    | 0.98                        


The count, mean, mininum, and max rows are self-explanatory. The StdDev row shows the standard deviation, which measures how dispersed the values are. The 25%, 50%, and 75% rows show the corresponding percentiles: a percentile indicates the value below which a given percentage of observations in a group of observations falls.

For example, 25% of the districts have a housing_median_age lower than
18, while 50% are lower than 29 and 75% are lower than 37. These are often called the
25th percentile (or 1st quartile), the median, and the 75th percentile (or 3rd quartile).

Another quick way to get a feel of the type of data you are dealing with is to plot a
histogram for each numerical attribute.

In [22]:
import DataFrame ((|>))

-- We can explicitly filter out numeric columns. `selectBy` lets us filter columns by
-- any of the below:
--    * column name
--    * column index range
--    * column property (isNumeric, isOptional) etc.
--    * column name range (similar to dplyr)

-- plotAllHistograms already skips non-numeric columns using similar logic.
-- We include this API here purely for demonstration.
Plt.plotAllHistograms (df |> D.selectBy [D.byProperty D.isNumeric])

<canvas id="chart_4Kjjtui7sCgZy2bdFcfL9GANAeAsUm1QiOmxr88ICegreF5BoLMC" style="width:100%;max-width:600px;height:400px"></canvas>
<script src="https://cdnjs.cloudflare.com/ajax/libs/Chart.js/2.9.4/Chart.js"></script>
<script>
new Chart("chart_4Kjjtui7sCgZy2bdFcfL9GANAeAsUm1QiOmxr88ICegreF5BoLMC", {
  type: "bar",
  data: {
    labels: ["1","204","406","609","812","1014","1217","1420","1623","1825","2028","2231","2433","2636","2839","3042","3244","3447","3650","3852","4055","4258","4460","4663","4866","5068","5271","5474","5677","5879"],
    datasets: [{
      label: "households",
      data: [2518,7695,5368,2495,1080,569,316,218,116,69,49,41,23,24,12,9,9,5,5,5,3,2,2,2,2,1,1,0,0,0],
      backgroundColor: "rgba(75, 192, 192, 0.6)",
      borderColor: "rgba(75, 192, 192, 1)",
      borderWidth: 1
    }]
  },
  options: {
    title: { display: true, text: "Histogram of households" },
    scales: {
      yAxes: [{ ticks: { beginAtZero: true } }]
    }
  }
});
</script>

<canvas id="chart_P6jSXzzLePyhmbAzLuzC7OrjopvFy1n1KrcF9J2xFXtLKM2" style="width:100%;max-width:600px;height:400px"></canvas>
<script src="https://cdnjs.cloudflare.com/ajax/libs/Chart.js/2.9.4/Chart.js"></script>
<script>
new Chart("chart_P6jSXzzLePyhmbAzLuzC7OrjopvFy1n1KrcF9J2xFXtLKM2", {
  type: "bar",
  data: {
    labels: ["1","3","4","6","8","10","11","13","15","16","18","20","21","23","25","26","28","30","32","33","35","37","38","40","42","44","45","47","49","50"],
    datasets: [{
      label: "housing_median_age",
      data: [62,253,404,175,411,518,238,714,1283,698,1072,911,847,478,1185,959,461,934,1180,689,1686,931,673,296,721,650,245,375,270,48],
      backgroundColor: "rgba(75, 192, 192, 0.6)",
      borderColor: "rgba(75, 192, 192, 1)",
      borderWidth: 1
    }]
  },
  options: {
    title: { display: true, text: "Histogram of housing_median_age" },
    scales: {
      yAxes: [{ ticks: { beginAtZero: true } }]
    }
  }
});
</script>

<canvas id="chart_IHJWbTGHAaiTKVUY8pRRvrD22Mb2ygZRh44C3QwppCjvpAwZZP2f" style="width:100%;max-width:600px;height:400px"></canvas>
<script src="https://cdnjs.cloudflare.com/ajax/libs/Chart.js/2.9.4/Chart.js"></script>
<script>
new Chart("chart_IHJWbTGHAaiTKVUY8pRRvrD22Mb2ygZRh44C3QwppCjvpAwZZP2f", {
  type: "bar",
  data: {
    labels: ["33","33","33","33","34","34","34","35","35","35","36","36","36","37","37","37","38","38","38","38","39","39","39","40","40","40","41","41","41","42"],
    datasets: [{
      label: "latitude",
      data: [1144,392,260,1480,5172,2335,374,143,230,231,53,190,324,596,372,1388,2139,1185,591,880,269,263,139,88,65,151,108,26,18,34],
      backgroundColor: "rgba(75, 192, 192, 0.6)",
      borderColor: "rgba(75, 192, 192, 1)",
      borderWidth: 1
    }]
  },
  options: {
    title: { display: true, text: "Histogram of latitude" },
    scales: {
      yAxes: [{ ticks: { beginAtZero: true } }]
    }
  }
});
</script>

<canvas id="chart_qOWa5NJaIlx3YIvUNBobEFfXUETsXHAA7ShLF9cPNzqKCkvGxs" style="width:100%;max-width:600px;height:400px"></canvas>
<script src="https://cdnjs.cloudflare.com/ajax/libs/Chart.js/2.9.4/Chart.js"></script>
<script>
new Chart("chart_qOWa5NJaIlx3YIvUNBobEFfXUETsXHAA7ShLF9cPNzqKCkvGxs", {
  type: "bar",
  data: {
    labels: ["-124","-124","-124","-123","-123","-123","-122","-122","-122","-121","-121","-121","-120","-120","-120","-119","-119","-119","-118","-118","-118","-117","-117","-117","-116","-116","-116","-115","-115","-115"],
    datasets: [{
      label: "longitude",
      data: [119,40,31,68,276,1209,2224,1432,908,836,445,392,227,610,261,580,313,1994,3595,1958,655,1876,266,110,69,19,95,3,8,20],
      backgroundColor: "rgba(75, 192, 192, 0.6)",
      borderColor: "rgba(75, 192, 192, 1)",
      borderWidth: 1
    }]
  },
  options: {
    title: { display: true, text: "Histogram of longitude" },
    scales: {
      yAxes: [{ ticks: { beginAtZero: true } }]
    }
  }
});
</script>

<canvas id="chart_n5X3vyx9GZ3wMBaRoeu9KAKv4xyZTc7xqA36kMKFz7kDxUrfC" style="width:100%;max-

One last thing you may want to do before actually preparing the data for Machine
Learning algorithms is to try out various attribute combinations. For example, the
total number of rooms in a district is not very useful if you don’t know how many
households there are. What you really want is the number of rooms per household.
Similarly, the total number of bedrooms by itself is not very useful: you probably
want to compare it to the number of rooms. And the population per household also
seems like an interesting attribute combination to look at.


We can reference existing attributes in type-safe expressions by automatically creating references to them. We use the `declareColumns` function from `DataFrame.Functions`. This creates global references to each of the columns.

In [6]:
:set -XTemplateHaskell -- The column generating function uses TemplateHaskell so we enable it first.
import qualified DataFrame.Functions as F
import qualified DataFrame.Display.Web.Plot as Plt

import Data.Text (Text)  -- So the compiler can refer to this type when generating columns

F.declareColumns df

In [7]:
priceBucket :: Double -> Text
priceBucket p
    | p > 500000 = ">500'000"
    | p > 450000 = ">450'000"
    | p > 400000 = ">400'000"
    | p > 300000 = ">300'000"
    | p > 200000 = ">200'000"
    | otherwise  = "<=200'000"

withPriceBuckets = df
                 |> D.derive "price_bucket" (F.lift priceBucket median_house_value)

-- We use the `name` function to get the text that our column refers to.
-- Saves us from having typos is the code.
Plt.plotScatterBy (F.name longitude) (F.name latitude) "price_bucket" withPriceBuckets

<canvas id="chart_apmZ7SJXvJiAQJiBv7ruLui2oSnTxzKDkmtPgK1lYYm" style="width:100%;max-width:600px;height:400px"></canvas>
<script src="https://cdnjs.cloudflare.com/ajax/libs/Chart.js/2.9.4/Chart.js"></script>
<script>
new Chart("chart_apmZ7SJXvJiAQJiBv7ruLui2oSnTxzKDkmtPgK1lYYm", {
  type: "scatter",
  data: {
    datasets: [
    {
      label: ">450'000",
      data: [{x:-122.23, y:37.88},{x:-122.18, y:37.81},{x:-122.25, y:37.86},{x:-122.25, y:37.86},{x:-122.23, y:37.82},{x:-122.24, y:37.83},{x:-122.24, y:37.83},{x:-121.92, y:37.53},{x:-121.88, y:37.64},{x:-122.04, y:37.85},{x:-121.93, y:37.76},{x:-122.03, y:37.86},{x:-121.95, y:37.81},{x:-122.12, y:37.91},{x:-122.12, y:37.85},{x:-122.22, y:37.88},{x:-122.18, y:37.89},{x:-121.82, y:37.81},{x:-118.52, y:34.29},{x:-118.52, y:34.3},{x:-118.51, y:34.24},{x:-118.51, y:34.23},{x:-118.52, y:34.23},{x:-118.51, y:34.23},{x:-118.41, y:34.16},{x:-118.58, y:34.15},{x:-118.59, y:34.14},{x:-118.56, y:34.18},{x:-118.43, y:34.15},{x:-118.45, y:34.15},{x:-118.47, y:34.15},{x:-118.48, y:34.15},{x:-118.45, y:34.14},{x:-118.36, y:34.13},{x:-118.38, y:34.14},{x:-118.41, y:34.14},{x:-118.28, y:34.11},{x:-118.31, y:34.13},{x:-118.34, y:34.11},{x:-118.3, y:34.06},{x:-118.35, y:34.08},{x:-118.36, y:34.07},{x:-118.37, y:34.07},{x:-118.37, y:34.07},{x:-118.37, y:34.07},{x:-118.37, y:34.07},{x:-118.38, y:34.08},{x:-118.38, y:34.06},{x:-118.37, y:34.06},{x:-118.39, y:34.05},{x:-118.31, y:34.05},{x:-118.45, y:34.06},{x:-118.44, y:34.05},{x:-118.43, y:34.05},{x:-118.43, y:34.04},{x:-118.43, y:34.04},{x:-118.39, y:34.05},{x:-118.39, y:34.04},{x:-118.43, y:34.01},{x:-118.44, y:34.02},{x:-118.41, y:34.03},{x:-118.42, y:34.0},{x:-118.46, y:33.98},{x:-118.4, y:33.98},{x:-118.4, y:33.97},{x:-118.29, y:33.71},{x:-118.27, y:34.22},{x:-118.23, y:34.18},{x:-118.21, y:34.17},{x:-118.27, y:34.18},{x:-118.34, y:34.15},{x:-117.83, y:34.15},{x:-117.87, y:34.15},{x:-117.85, y:34.06},{x:-117.88, y:34.06},{x:-118.03, y:34.15},{x:-118.04, y:34.18},{x:-118.03, y:34.16},{x:-118.05, y:34.12},{x:-118.07, y:34.17},{x:-118.11, y:34.19},{x:-118.19, y:34.2},{x:-118.21, y:34.21},{x:-118.08, y:34.14},{x:-118.12, y:34.14},{x:-118.13, y:34.13},{x:-118.18, y:34.13},{x:-118.14, y:34.13},{x:-118.15, y:34.11},{x:-118.15, y:34.12},{x:-118.17, y:34.1},{x:-117.99, y:33.95},{x:-118.0, y:33.96},{x:-118.1, y:33.94},{x:-118.11, y:33.95},{x:-118.15, y:33.84},{x:-118.2, y:33.83},{x:-118.13, y:33.79},{x:-118.16, y:33.72},{x:-118.14, y:33.77},{x:-118.12, y:33.76},{x:-118.12, y:33.77},{x:-118.4, y:33.9},{x:-118.41, y:33.9},{x:-118.4, y:33.9},{x:-118.4, y:33.89},{x:-118.39, y:33.89},{x:-118.38, y:33.88},{x:-118.38, y:33.88},{x:-118.39, y:33.88},{x:-118.39, y:33.88},{x:-118.39, y:33.88},{x:-118.39, y:33.88},{x:-118.4, y:33.88},{x:-118.4, y:33.87},{x:-118.4, y:33.87},{x:-118.37, y:33.81},{x:-118.38, y:33.81},{x:-118.37, y:33.81},{x:-118.38, y:33.81},{x:-118.39, y:33.81},{x:-118.5, y:34.02},{x:-118.49, y:34.02},{x:-118.49, y:34.02},{x:-118.48, y:34.0},{x:-118.47, y:34.01},{x:-118.47, y:34.01},{x:-118.47, y:34.01},{x:-118.47, y:34.01},{x:-118.45, y:34.02},{x:-118.46, y:34.02},{x:-118.46, y:34.02},{x:-118.39, y:34.0},{x:-118.35, y:34.0},{x:-118.6, y:34.13},{x:-122.51, y:37.98},{x:-122.49, y:37.98},{x:-122.52, y:37.95},{x:-122.53, y:37.96},{x:-122.5, y:37.88},{x:-122.51, y:37.89},{x:-122.54, y:37.91},{x:-122.55, y:37.9},{x:-122.56, y:37.9},{x:-122.55, y:37.91},{x:-122.62, y:37.85},{x:-122.53, y:37.86},{x:-122.71, y:37.88},{x:-121.92, y:36.54},{x:-121.94, y:36.55},{x:-121.92, y:36.55},{x:-121.92, y:36.55},{x:-121.95, y:36.6},{x:-121.94, y:36.58},{x:-121.89, y:36.6},{x:-117.89, y:33.89},{x:-117.94, y:33.89},{x:-117.81, y:33.88},{x:-117.82, y:33.84},{x:-117.81, y:33.84},{x:-117.62, y:33.64},{x:-117.58, y:33.6},{x:-117.53, y:33.69},{x:-117.68, y:33.51},{x:-117.71, y:33.51},{x:-117.71, y:33.54},{x:-117.74, y:33.46},{x:-117.79, y:33.55},{x:-117.77, y:33.55},{x:-117.76, y:33.54},{x:-117.8, y:33.64},{x:-117.8, y:33.63},{x:-117.87, y:33.6},{x:-117.87, y:33.6},{x:-117.8

From the plot above we can see that there are specific population centers where house prices are high. This tells us that a really good model should be able to account for this non-linearity. The expensive population centers are coastal suggesting that the `ocean_proximity` will be useful for predictiting house prices.

Now let's declare these attributes.

In [17]:
import Data.Maybe (fromMaybe)

ordinalOceanProximity :: Text -> Double
ordinalOceanProximity op = case op of
    "ISLAND"     -> 5
    "NEAR OCEAN" -> 4
    "NEAR BAY"   -> 3
    "<1H OCEAN"  -> 2
    "INLAND"     -> 1
    _ -> error ("Unknown ocean proximity value: " ++ show op)

augmented = df D.|> D.derive "rooms_per_household" (total_rooms / households)
               D.|> D.derive "bedrooms_per_household" (F.lift (fromMaybe 0) total_bedrooms / total_rooms)
               D.|> D.derive "population_per_household" (population / households)
               D.|> D.derive "ocean_proximity" (F.lift ordinalOceanProximity ocean_proximity)

In [9]:
import           Control.Monad
import qualified Data.List as L
import           Data.Function

In [18]:
correlationWithHouseValue columnName = (columnName, fromMaybe 0 (D.correlation columnName "median_house_value" augmented))

correlations = map correlationWithHouseValue (D.columnNames augmented)

mapM_ print (L.sortBy (flip compare `on` snd) correlations)

("median_house_value",1.0)
("median_income",0.6880752079585185)
("ocean_proximity",0.38324444779865596)
("rooms_per_household",0.1519482897414869)
("total_rooms",0.13415311380656322)
("housing_median_age",0.10562341249320993)
("households",6.584265057005642e-2)
("total_bedrooms",0.0)
("population_per_household",-2.373741295613557e-2)
("population",-2.4649678888894886e-2)
("longitude",-4.5966615119112986e-2)
("latitude",-0.14416027687448565)
("bedrooms_per_household",-0.23875927633089492)

Hey, not bad! The new bedrooms_per_room attribute is much more correlated with
the median house value than the total number of rooms or bedrooms. Apparently
houses with a lower bedroom/room ratio tend to be more expensive. The number of
rooms per household is also more informative than the total number of rooms in a
district—obviously the larger the houses, the more expensive they are.

Also, once we translate ocean proximity to an ordinal value 5 means close the ocean and 1 means far from the ocean we see how predictive the features is.